hacspec/src/seq.rs

```diff
@@ -2,372 +2,317 @@
-    }
-    /// Update this sequence with `v` at position `start_out`.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use hacspec::prelude::*;
-    ///
-    /// let mut s = Seq::<u8>::new(5);
-    /// s = s.update_element(4, 7);
-    /// assert_eq!(s, Seq::<u8>::from_array(&[0, 0, 0, 0, 7]));
-    /// ```
-    pub fn update_element(mut self, start_out: usize, v: T) -> Self {
-        debug_assert!(self.len() >= start_out + 1);
-        self[start_out] = v;
-        self
-    }
-    /// Reset the sequence index.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use hacspec::prelude::*;
-    ///
-    /// let mut s = Seq::<u8>::new(5);
-    /// s = s.set_index(3);
-    /// s = s.push(Seq::<u8>::from_array(&[4, 5]));
-    /// assert_eq!(s, Seq::<u8>::from_array(&[0, 0, 0, 4, 5]));
-    /// ```
-    pub fn set_index(self, i: usize) -> Self {
-        Self {
-            b: self.b.clone(),
-            idx: i,
-        }
-    }
-    /// Push `v` to this sequence and move `idx` according to `v.len()`.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use hacspec::prelude::*;
-    ///
-    /// let mut s = Seq::<u8>::new(5);
-    /// s = s.set_index(3);
-    /// s = s.push(Seq::<u8>::from_array(&[4, 5]));
-    /// assert_eq!(s, Seq::<u8>::from_array(&[0, 0, 0, 4, 5]));
-    /// ```
-    pub fn push<A: SeqTrait<T>>(self, v: A) -> Self {
-        println!("{:?} >= {:?} + {:?}", self.len(), self.idx, v.len());
-        debug_assert!(self.len() >= self.idx + v.len());
-        let idx = self.idx;
-        let mut self_copy = self;
-        for (i, b) in v.iter().enumerate() {
-            self_copy[idx + i] = *b;
-        }
-        self_copy.idx += v.len();
-        self_copy
-    }
-    /// Push `l` elements from `v` to this sequence and move `idx` according to `l`.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use hacspec::prelude::*;
-    ///
-    /// let mut s = Seq::<u8>::new(5);
-    /// s = s.set_index(3);
-    /// s = s.push_sub(Seq::<u8>::from_array(&[4, 5]), 1, 1);
-    /// assert_eq!(s, Seq::<u8>::from_array(&[0, 0, 0, 5, 0]));
-    /// ```
-    pub fn push_sub<A: SeqTrait<T>>(self, v: A, start: usize, l: usize) -> Self {
-        debug_assert!(self.len() >= self.idx + l);
-        debug_assert!(v.len() >= start + l);
-        let idx = self.idx;
-        let mut self_copy = self;
-        for (i, b) in v.iter().skip(start).take(l).enumerate() {
-            self_copy[idx + i] = *b;
+        impl<T: $bound $(+ $others)*> $name<T> {
+            #[primitive(hacspec)]
+            pub fn new(l: usize) -> Self {
+                Self {
+                    b: vec![T::default(); l],
+                }
+            }
+
+            /// Get the size of this sequence.
+            #[primitive(hacspec)]
+            pub fn len(&self) -> usize {
+                self.b.len()
+            }
+
+            #[library(hacspec)]
+            pub fn sub(self, start_out: usize, len: usize) -> Self {
```

Comment:
Yes but no, because there are slight differences in the API of these functions depending on whether they're `PublicSeq` or `SecretSeq`. Indeed, we want `sub` on a `SecretSeq` to return a `SecretSeq` and not a generic `Seq` right ? We cannot have that if we put it inside `SeqTrait`.